spec/component/forms/new_deputy_assignment_form_spec.rb

```diff
@@ -0,0 +1,241 @@
+
+  before do
+    allow(PsuIdentity::SearchService::Client).to receive(:new).and_return(psu_identity_client)
+  end
+
+  describe 'validations' do
+    it { is_expected.to validate_presence_of(:primary) }
+    it { is_expected.to validate_presence_of(:deputy_webaccess_id) }
+
+    it 'validates the basic format of deputy_webaccess_id' do
+      expect(form).to allow_value('abc123').for(:deputy_webaccess_id)
+      expect(form).to allow_value('ABC123').for(:deputy_webaccess_id)
+
+      expect(form).not_to allow_value('abc123@psu.edu').for(:deputy_webaccess_id)
+      expect(form).not_to allow_value('abc 123').for(:deputy_webaccess_id)
+    end
+  end
+
+  describe '#save' do
+    context 'when no User exists for the given webaccess id' do
+      let(:psu_identity_data) {
+        { 'userid' => 'dep0987', 'cprid' => '123456', 'givenName' => 'Deputy', 'familyName' =>
+'FromPsu', 'active' => true, 'confHold' => false, 'serviceAccount' => false, "affiliati
+on" => ['MEMBER'], 'displayName' => 'Deputy FromPsu', 'link' => { 'href' => "https://dev.app
+s.psu.edu/cpr/resources/123456" } }
+      }
+      let(:psu_person) { PsuIdentity::SearchService::Person.new(psu_identity_data) }
+
+      context 'when PsuIdentity responds with a valid response' do
+        before { allow(psu_identity_client).to receive(:userid).with(deputy_webaccess_id).and_return(psu_person) }
+
+        it 'returns true' do
+          expect(form.save).to eq true
+        end
+
+        it 'creates a User' do
+          expect {
+            form.save
+          }.to change(User, :count).by(1)
+
+          deputy = User.find_by(webaccess_id: deputy_webaccess_id)
+          expect(deputy).to be_present
+          expect(deputy.first_name).to eq 'Deputy'
+          expect(deputy.last_name).to eq 'FromPsu'
+          expect(deputy.psu_identity).to be_present
+        end
+
+        it 'creates an active DeputyAssignment' do
+          expect {
+            form.save
+          }.to change(DeputyAssignment.active, :count).by(1)
+
+          assignment = DeputyAssignment.active.last
+          expect(assignment.primary).to eq primary
+          expect(assignment.deputy.webaccess_id).to eq deputy_webaccess_id
+        end
+      end
+
+      context 'when PsuIdentity raises an error' do
+        before { allow(psu_identity_client).to receive(:userid).and_raise(StandardError) }
+
+        it 'returns false' do
+          expect(form.save).to eq false
+        end
+
+        it 'has a nice error' do
+          form.save
+          expect(form.errors.messages_for(:base)).to include i18n_error('base.identity_service_error')
+        end
+
+        it 'creates nothing' do
+          expect {
+            form.save
+          }.to not_change(User, :count)
+            .and not_change(DeputyAssignment, :count)
+        end
+      end
+
+      context 'when PsuIdentity responds with something, but the data is invalid' do
```

Comment:
I don't know when it would happen, but given that this code has two external dependencies (1. the validations on User, 2. whatever is returned by the identity server) which may not match each other, it seems like a case worth handling

Reply:
Real life example of this happening from Dan C: if someone has a "confidentiality hold" then they will not have a first/last name in the PSU identity server response. RMD requires a User to have both a first and last name.